controller.py

```diff
@@ -266,16 +249,50 @@ def libraries(self, live=True):
 
         data = dict(libraries=result)
         now = datetime.datetime.utcnow()
-        self.library_list_cache[live] = (data, now)
         return data
 
-    def libraries_opds(self, live=True):
-        """Return all the libraries in OPDS format"""
+    def libraries_opds(self, live=True, location=None):
+        """Return all the libraries in OPDS format
+
+        :param live: If this is True, then only production libraries are shown.
+        :param location: If this is set, then libraries near this point will be
+           promoted out of the alphabetical list.
+        """
+        alphabetical = self._db.query(Library).order_by(Library.name)
 
-        libraries = self._db.query(Library).order_by(Library.name)
         # We always want to filter out cancelled libraries.  If live, we also filter out
         # libraries that are in the testing stage, i.e. only show production libraries.
-        libraries = libraries.filter(Library._feed_restriction(production=live))
+        alphabetical = alphabetical.filter(Library._feed_restriction(production=live))
+
+        # Pick up each library's hyperlinks and validation
+        # information; this will save database queries when building
+        # the feed.
+        alphabetical = alphabetical.options(
+            joinedload('hyperlinks'),
+            joinedload('hyperlinks', 'resource'),
+            joinedload('hyperlinks', 'resource', 'validation'),
+        )
+        if location is None:
+            # No location data is available. Use the alphabetical list as
+            # the list of libraries.
+            libraries = alphabetical
+        else:
+            # Location data is available. Get the list of nearby libraries, then get
+            # the rest of the list in alphabetical order.
+
+            # We can't easily do the joindeload() thing for this
+            # query, because it doesn't simply return Library objects,
+            # but it won't return more than five results.
+            nearby_libraries = Library.nearby(
+                self._db, location, production=live
+            ).limit(5).all()
+
+            # Exclude nearby libraries from the alphabetical query
+            # to get a list of faraway libraries.
+            faraway_libraries = alphabetical.filter(
+                ~Library.id.in_([x.id for x, distance in nearby_libraries])
```

Comment:
Are you excluding the nearby libraries by returning the inverse of the nearby libraries?

Reply:
I wouldn't put it like that. The `~Library.id.in_` syntax here gets translated to the SQL syntax `libraries.id NOT IN`.

Reply:
Ah, that's because this is still a SQL object, right?

Reply:
Right, every call to `filter` is adding another WHERE clause to the eventual query.